controllers/keystoneapi_controller.go

```diff
@@ -109,6 +109,25 @@ func (r *KeystoneAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, err
 	}
 
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If we're not deleting this and the service object doesn't have our finalizer, add it.
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
+		// Register the finalizer immediately to avoid orphaning resources on delete
+		err := r.Update(ctx, instance)
+
+		return ctrl.Result{}, err
```

Comment:
If we have Update + return, then we can simply have just a return here and let the deferred call do the Update. This would limit the places we do Updates on ourselves. See nova-controller as an example:
- https://github.com/openstack-k8s-operators/nova-operator/blob/fc2fe084dc1031a0d6582a14f59562f88fc70305/controllers/nova_controller.go#L109-L149
- https://github.com/openstack-k8s-operators/nova-operator/blob/fc2fe084dc1031a0d6582a14f59562f88fc70305/controllers/nova_controller.go#L159-L167

Reply:
Yes, we should let the `defer()` update do the actual call as much as possible.  I will try that out here.